new/multiple-policy-revision.md

```diff
@@ -0,0 +1,110 @@
+
+
+## Specification
+
+The proposed solution to the motivation described above is to allow a more composable form of policy which would be able to optionally include other policies inside it.
+
+This could be done with the addition of a ```include_policy``` directive, an example of which is shown below:
+
+```ruby
+include_policy "base", git: "github.com/myorg/policies.git", path: "foo/bar/baz.lock.json"
+```
+
+The ```include_policy``` directive will support three sources for policies: 
+
+* ```git```, with the following parameter being the URL to a Git repository and a path parameter specifying the location of the file within the repository
+* ```server```, with the following paramater being set to a Chef server URL, a policy_name paramter and a policy_revision_id parameter.
+* ```local```, with the following parameter being a path to a file on disk.
+
+When the ```chef update``` command is used to apply any changes to a policyfile containing the ```include_policy``` directive, any cookbook locks from the lockfile of the included policyfile will be pulled into the parent policy before its own .lock file is computed.
+
+When included policies come from a ```git``` source, the SHA of the commit at the time the included lockfile was first pulled into the parent will be stored in the parent lockfile and used when the included Lockfile must be reprocessed. This ensures that only the policyfile for which the update command was called has changed. Similarly, when included policies come from a ```server``` source, the revision id of the included lockfile will be stored in the parent lockfile. In the event of policy files being included from a ```local``` source, this guarantee cannot be given and the latest Lockfile for the included policy will be used.
+
+Essentially what this means is that the parent .lock file is computed from the merging of the following:
+
+* Data contained in the parent policyfile's .rb file
+* Computed cookbook locks from the lock files for all policies specified with an ```include_policy``` directive.
+
+The single fused lockfile produced by the above would then be uploaded to the Chef server as normal, and would function as a single Policy.
+
+
+## Problems
+
+The principal potential issue with this approach is that because we are pulling cookbook locks from a number of included cookbooks, it is necessary for all included policy Lockfiles to be re-scanned upon regeneration of the parent Policyfile, so that the full set of cookbook locks and constraints can be examined and merged. Although using the ```git``` source to include Policies will mean that the same commit SHA is used every time to rescan the included Lockfile and the ```server``` source will always ensure the same revision ID is used, when the ```local``` source is used we have to use whatever the current Lockfile present on disk is. This means that we cannot guarantee it has not changed since the last time it was scanned.
+
```

Comment:
I think there are two major issues with trying to squash multiple policy files into a single ```parent```

1) Attribute overrides - If two included policy files set the same attribute how is this handled? It almost has to be an error otherwise policy files can override each others attributes and the results would be almost chaos. Debugging this situation as a developer of an included policy file which worked everywhere except where it was included would be difficult at best. 

2) In the context where all the included Policyfiles are squashed and then compiled as a single Policyfile does this mean that the computed locks for all common dependencies have to be identical? Otherwise the behaviour would be the contrary to the intention of Policyfiles where dependencies could change depending on where the Policyfile was included.